JavaScriptでメモの一覧・参照・追加・編集・削除ができるコマンドラインアプリを作成する

nakamu-kazu222 opened the pull request:
以下の機能を実装したメモコマンドラインアプリを作成する
- 追加
  - 標準入力に入ってきたテキストを新しいメモとして追加する
    ```
    $ echo 'メモの内容' | memo.js
    ```
- 一覧
  - それぞれのメモの最初の行のみを表示する
    ```
    $ memo.js -l
    メモ1
    今日の日記
    晩ご飯のレシピ
    ```

- 参照
  - 選んだメモの全文が表示される
    - カーソルを合わせた時点もしくはメモを選択した時にメモの全文が表示される
      ```
      $ memo.js -r
      Choose a note you want to see:
        メモ1
        今日の日記
      > 晩ご飯のレシピ
        
        晩ご飯のレシピ
        カレー
        豚肉
        じゃがいも
        人参
        タマネギ
        カレールー
      ```

- 削除
  - 選んだメモが削除される
    ```
    $ memo.js -d
    Choose a note you want to delete:
      メモ1
      今日の日記
    > 晩ご飯のレシピ
    ```

- 保存先ストレージ
  - memoの保存先はsqlite3で実装する

nakamu-kazu222 committed:
ESLint と Prettier でチェックできるようにした

nakamu-kazu222 committed:
minitestをインストールした

nakamu-kazu222 committed:
enquirerをインストールした

nakamu-kazu222 committed:
Enquirer ライブラリを使用して、コンソールからの入力と出力を制御するクラスを作成した

nakamu-kazu222 committed:
コマンドラインの入力を処理し、メモの追加、リスト表示、参照または削除の操作を実行するクラスを作成した

nakamu-kazu222 committed:
メモの作成、追加、リスト表示、削除のデータベースの操作を処理するクラスを作成した

nakamu-kazu222 committed:
メモアプリを実行するファイルを作成した

yuuu commented on 04.class/memo_database.js:
```diff
@@ -0,0 +1,29 @@
+  }
+
+  list_memos(callback) {
+    this.db.all("SELECT * FROM memos", callback);
```
DBによっては順序がランダムになる可能性があるので、明示的に `order` をしておいたほうが良いですね。

nakamu-kazu222 replied:
order で降順になるようにしました

yuuu commented on 04.class/memo_command.js:
```diff
@@ -0,0 +1,117 @@
+    this.command = this.option[0];
+  }
+
+  async process_command() {
```
このメソッドは `Promise` を返しますが、呼び元では `await` ができないので、実際には非同期処理が完了する前に呼び元が終了する可能性があります。

最近のNode.jsはTop-Level Awaitに対応しているはずなので、呼び元に `await` を付けるか、難しそうであればこのメソッドの `async` を外して `then` で処理するようにしましょう。

nakamu-kazu222 replied:
呼び出し元にawaitをつけ、process_command() メソッドの処理が完了するまで待機するようにしました

yuuu commented on 04.class/memo_command.js:
```diff
@@ -0,0 +1,117 @@
+    });
+  }
+
+  async view_or_delete_option() {
```
viewとdeleteを同じメソッドで処理しようとしていることで、複数箇所で `this.command` を使った分岐が発生していますね。

viewとdeleteのメソッド自体は分けて、共通する処理をメソッドに切り出す方針として複雑度を下げられないか検討をお願い致します 🙏

nakamu-kazu222 replied:
viewとdeleteのメソッドを分け、共通の処理を別のメソッドに分けました

nakamu-kazu222 committed:
order で降順になるようにした

nakamu-kazu222 committed:
lintで全てのファイルをチェックでき、不要なので消去した

nakamu-kazu222 committed:
非同期処理が完了するまで待機するようにした

process_command()メソッドはPromiseを返すので、処理が完了するまで待機しなければならない
そのため、呼び出し元で待機するようにした

nakamu-kazu222 committed:
viewとdeleteのメソッドを分け、共通の処理を別のメソッドに分けた

複数箇所で this.command を使った分岐が発生しており、冗長だった

yuuu commented on 04.class/memo_command.js:
```diff
@@ -0,0 +1,122 @@
+            choices
+          );
+
+          const selected_memo = memos.find(
```
`enquirer` から返された値には `id` を含められるはずなので、そのIDを使ってDB側で絞り込みをした方が効率的ですね。

yuuu replied:
delete操作も `id` が取得できていれば当該メモを探し直さずとも `DELETE` のクエリが発行できるのでこのあたりの処理は省略できそうです。

nakamu-kazu222 replied:
findメソッドの比較条件を`memo.id === answer.selected_item`にして、選択されたメモのID (answer.selected_item) と一致する memo.id を取得するようにしました

nakamu-kazu222 replied:
select_memo_listで、選択したメモのidを取得してクエリを発行するようにしました

nakamu-kazu222 committed:
選択したアイテムのidを取得できるようにした

nakamu-kazu222 committed:
メモを表示するSQLを追加した

nakamu-kazu222 committed:
選択したメモのidを取得してクエリを発行するようにした

get_memoで選択したメモを表示できるようにした
view_optionとdelete_optionのSQLのidを選択したメモのidを取得できるようにした
list_option()、view_option()、delete_option()の処理を共通化した

yuuu commented on 04.class/console_operator.js:
```diff
@@ -0,0 +1,28 @@
+      message,
+      choices,
+      result() {
+        return this.focused.value;
```
こちら、もっと簡単な方法があると思っていましたが、実際にはこの方法しかないんですね 💦 対応ありがとうございます 🙏
